Scripts/Install-PatchOrchestrator.ps1

```diff
@@ -23,46 +34,56 @@ catch
-
-$action = 'install'
-$installedPackage = C:\choco\choco list -lo | Where-object { $_.ToLower().Contains('servicefabric-patchorchestrator') }
-$install = $true
-
-if ($installedPackage.Count -gt 0)
-{
-    $action = 'upgrade'
-    $nameVersion = $installedPackage -split ' '
+# Set POA Application Parameters
+$PoaParameters = @{
+    WUFrequency = "$PoaUpdateFrequency, $($PoaUpdateTime)"
+    TaskApprovalPolicy = "$PoaApprovalPolicy"
+    MinWaitTimeBetweenNodes = "$PoaWaitTimeBetweenNodes"
+}
 
-    if ($nameVersion[1] -eq $Version)
-    {
-        $install = $false
-    }
+# Check if POA version has been set. If not, use latest version.
+if ($PoaVersion -eq "latest") {
+    $PoaRelease = (Invoke-RestMethod "https://api.github.com/repos/microsoft/Service-Fabric-POA/releases/latest" -UseBasicParsing)
+} else {
+    $PoaRelease = (Invoke-RestMethod "https://api.github.com/repos/microsoft/Service-Fabric-POA/releases" -UseBasicParsing) | Where-Object {$_.tag_name -eq "v$($PoaVersion)"}
 }
 
-if ($install)
-{
-    C:\choco\choco $action servicefabric-patchorchestrator -s "'$packageDir'" --version "'$Version'" --confirm --pre --allow-downgrade --timeout 3600
+$PoaVersion = $PoaRelease.tag_name -replace ("v","")
+$PoaReleaseDownload = $PoaRelease.assets.browser_download_url | Where-Object {$_ -like "*.zip"}
 
-    if (!$?)
-    {
-        Write-Error "Failed to install package servicefabric-patchorchestrator"
+if ($PoaReleaseDownload) {
+    # Download and unpack POA package from GitHub
+    Invoke-RestMethod -Uri $PoaReleaseDownload -UseBasicParsing -OutFile "$TempDirectory\POA.zip"
+    Expand-Archive "$TempDirectory\POA.zip" -DestinationPath "$TempDirectory\POA" -Force
+    Set-Location "$TempDirectory\POA"
+    
+    # Check for App and Install or Upgrade
+    $PoaApp = Get-ServiceFabricApplication fabric:/PatchOrchestrationApplication
+    if ($PoaApp) {
+        $PoaAppUpgradeStatus = Get-ServiceFabricApplicationUpgrade -ApplicationName fabric:/PatchOrchestrationApplication
+        if ($PoaAppUpgradeStatus.UpgradeDomainsStatus.State -notmatch "Completed") {
+            Write-Host "POA application install already in progress"
+            Exit 0
+        } else {
+            if ($PoaApp.ApplicationTypeVersion -eq "$PoaVersion") {
+                # Upgrade POA settings
+                Start-ServiceFabricApplicationUpgrade $PoaApp.ApplicationName -ApplicationTypeVersion $PoaApp.ApplicationTypeVersion -FailureAction Rollback -Monitored -ApplicationParameter $PoaParameters
+            } else {
+                # Upgrade POA Package
+                .\Upgrade.ps1 -ApplicationParameters $PoaParameters
+            }
+        }
+    } else {
+        # Deploy POA package
+        .\Deploy.ps1 -ApplicationParameters $PoaParameters
     }
-
-    $oldPackages | Remove-Item -Force
+} else {
+    # Exit on error
+    Write-Host "[ERROR] Failed to find Release for version $PoaVersion" -ForegroundColor Red

```

Comment:
Do we get errors anywhere when they occur?

Reply:
That specific line is just an `else` for a where condition so no error. I didn't handle errors in this one because the script is set to stop on error and the exception is visible natively in Azure portal where the runbook is executed.